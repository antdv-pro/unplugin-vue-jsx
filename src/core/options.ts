import process from 'node:process'
import { getPackageInfoSync } from 'local-pkg'
import type { VueJSXPluginOptions } from '@vue/babel-plugin-jsx'
import type { FilterPattern } from 'unplugin-utils'

export type Vue2JSXOptions = {
  functional?: boolean
  injectH?: boolean
  vModel?: boolean
  vOn?: boolean
  compositionAPI?: 'auto' | 'native' | 'plugin' | boolean | 'naruto'
}

export type Options = {
  include?: FilterPattern
  exclude?: FilterPattern | undefined
  root?: string
  sourceMap?: boolean
} & (OptionsVue2 | OptionsVue3 | { version?: 'auto' })

interface OptionsVue2 extends Vue2JSXOptions {
  version: 2
}

interface OptionsVue3 extends VueJSXPluginOptions {
  version: 3
}

export type OptionsResolved = Omit<Required<Options>, 'version' | 'exclude'> & {
  version: 2 | 3
  exclude?: Options['exclude']
}

export function resolveOptions(options: Options): OptionsResolved {
  const root = options.root || process.cwd()
  let version: 2 | 3
  if (options.version === 'auto') {
    version = getVueVersion(root)
  } else version = options.version || 3

  return {
    ...options,
    include: options.include || [/\.[jt]sx?$/],
    exclude: options.exclude || [/node_modules/],
    version,
    root,
    sourceMap: options.sourceMap ?? true,
  }
}

function getVueVersion(root: string) {
  const pkg = getPackageInfoSync('vue', { paths: [root] })
  return pkg?.version?.startsWith('2') ? 2 : 3
}
