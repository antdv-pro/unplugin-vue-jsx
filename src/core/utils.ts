export const isTS = (id: string) => /\.tsx?$/.test(id)
