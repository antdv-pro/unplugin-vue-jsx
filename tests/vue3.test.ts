import { describe, expect, test } from 'vitest'
import { resolveOption } from '../src/core/options'
import { transformVue3 } from '../src/core/vue3'

const transform = async (code: string, isTS = false) => {
  const options = await resolveOption({ version: 3 })
  return (await transformVue3(code, `foo.${isTS ? 'tsx' : 'jsx'}`, options))
    ?.code
}

describe('Vue 3', async () => {
  test('basic', async () => {
    expect(await transform('<div />')).toMatchSnapshot()
    expect(await transform('<div key="1" />')).toMatchSnapshot()
    expect(await transform("<div foo={'bar'} />")).toMatchSnapshot()
  })

  test('typescript', async () => {
    expect(await transform('const foo: any = <div />', true)).toMatchSnapshot()
  })
})
