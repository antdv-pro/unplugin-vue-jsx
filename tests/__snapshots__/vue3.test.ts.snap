// Vitest Snapshot v1

exports[`Vue 3 > basic 1`] = `
"import { createVNode as _createVNode } from \\"vue\\";

_createVNode(\\"div\\", null, null);"
`;

exports[`Vue 3 > basic 2`] = `
"import { createVNode as _createVNode } from \\"vue\\";

_createVNode(\\"div\\", {
  \\"key\\": \\"1\\"
}, null);"
`;

exports[`Vue 3 > basic 3`] = `
"import { createVNode as _createVNode } from \\"vue\\";

_createVNode(\\"div\\", {
  \\"foo\\": 'bar'
}, null);"
`;

exports[`Vue 3 > typescript 1`] = `
"import { createVNode as _createVNode } from \\"vue\\";

const foo: any = _createVNode(\\"div\\", null, null);"
`;
